docs/hooks.md

```diff
@@ -59,7 +59,91 @@ This is used in SystemJS core to ensure that import maps are loaded so that the
+
+// Event Loop N                                                         // lastRegister is undefined
+[code.js]: System.register(module_code)                                 // lastRegister is module_code
+  => [system.js]: (Notified by a callback that code.js has executed)    // lastRegister is module_code
+  => [system.js]: let temp = getRegister()                              // lastRegister is undefined, temp is module_code
+  => [system.js]: Promise_ID_1 resolved by temp                         // lastRegister is undefined, "code.js" loaded
+```
+
+The **WRONG** execution order may look like: (ends with wrong module registered and [Module did not instantiate](https://github.com/systemjs/systemjs/blob/master/docs/errors.md#2) error).
+
+```js
+// Event Loop 0                                                     // lastRegister is undefined
+[app.js] System.import("./library.js")
+  => [system.js] System.instantiate("./library.js") => Promise_ID_0
+  => ...load library.js...
+
+// Event Loop 0                                                     // lastRegister is undefined
+[app.js] System.import("./code.js")
+  => [system.js] System.instantiate("./code.js") => Promise_ID_1
+  => ...load code.js...
+
+// Event Loop N                                                     // lastRegister is undefined
+[library.js]: System.register(module_library)                       // lastRegister is module_library
+[code.js]: System.register(module_code)                             // lastRegister is module_code !!! module_library has been overwrittened !!!
+
+// Event Loop M
+[system.js]: (Notified by a callback that library.js has executed)  // lastRegister is module_code
+  => [system.js]: let temp = getRegister()                          // lastRegister is undefined, temp is module_code
+  => [system.js]: Promise_ID_0 resolved by temp                     // lastRegister is undefined, !!! "library.js" loaded with content of "code.js" !!!
+
+// Event Loop X
+[system.js]: (Notified by a callback that code.js has executed)     // lastRegister is undefined
+  => [system.js]: let temp = getRegister()                          // lastRegister is undefined, temp is undefined
+  => [system.js]: throw "Module did not instantiate"
+```
+
+The [default implementation](https://github.com/systemjs/systemjs/blob/master/src/features/script-load.js) is based on the creation of `<script>` tag and the `onload` event is emitted just after the System.register call.
+
+If in your environment it is not possible to maintain the execution order (for example, involves the interaction of multiple event loops), you can make the call to the instantiate in a sequent way (Notice: this will slow down the module loading speed if they need to loaded by the network). Here is another [example that implementing System.instantiate](https://github.com/Jack-Works/webextension-systemjs/blob/master/src/content-script.ts#L10-L26) for a speical environment that cannot use `<script>` tag.
```

Comment:
Do you mean https://github.com/systemjs/systemjs/blob/853c4a85c4c814ca04bb2b65e13798cf23eee91c/src/features/node-fetch.js?

Reply:
oh it's https://github.com/systemjs/systemjs/blob/master/src/features/fetch-load.js my branch seems like outdated

Reply:
`fetch-load` is calling `getRegister` immediately after `eval` and not showing how to handle the race condition (cause it doesn't have race condition).

My [webextension-systemjs](https://github.com/Jack-Works/webextension-systemjs) involved this topic (cause I'm not using `eval` but `browser.tabs.executeScript` which is async and not reliably to get a callback of execution done).

Reply:
If you're not calling `getRegister` immediately after `eval` then you **will** have race conditions.

Reply:
It's impossible to know if the script executed synchronously in some environments. I do have race conditions in my early implementation of `webextension-systemjs` but I successfully workaround on it.
https://github.com/Jack-Works/webextension-systemjs/blob/master/src/content-script.ts#L10-L26, it chains the current promise after the last execution (to make it serial) therefore to maintain the correct order (and still keep asynchronous).

Reply:
Can you really not just call eval manually? Promise resolution runs the task queue, which means you always have a very small window where a race can still happen even if it hasn't happened in testing, it could be 1 in 100 or 1 in 10000 but the race is still there.

Reply:
I used it in our products directly, and the current version works well. `eval` is banned in our environment by CSP policy (banning `eval` is enabled by default in WebExtension).

Reply:
What's the reason you can't attach a sync callback in the page with eg a page registration queue handler that messages back both the registration and url together?

Reply:
Sorry, I didn't get your idea 👀 can you write some pseudo-code to explain?